### PR TITLE
Autofill presenter name with the info from /me.json

### PR DESCRIPTION
**Pull request by narickmann**
If the presenter field is empty (and no value in local storage) the users name from info/me.json is used.

### CODE REVIEW

**Review by lkiesow — CHANGES_REQUESTED** [review]
If Opencast is unreachable, Studio shows an error. Once the connection can be established, the error disappears and the event can be uploaded. This code seems to cause a stack trace and seems to be unable to recover, however:

```
error when getting info/me Error: network error when accessing 'https://develop.opencast.org/info/me.json': TypeError: NetworkError when attempting to fetch resource.
    q https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    Z https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    request https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    jsonRequest https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    getInfoMe https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    updateUser https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    refreshConnection https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    e https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/823.1e6ee22f.chunk.js:1
    setTimeout handler*7448/Ae/< https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/823.1e6ee22f.chunk.js:1
    nl https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    Su https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    ku https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    S https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    A https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    6813 https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    __webpack_require__ https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    5296 https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    __webpack_require__ https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    4463 https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    __webpack_require__ https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    4164 https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    __webpack_require__ https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    <anonymous> https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    <anonymous> https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
    <anonymous> https://test.studio.opencast.org/2023-05-11_08-04-44-pr1037-4945454464/static/js/main.ce2c9eee.js:2
main.ce2c9eee.js:2:7963
```

This also seems to cause studio to load `/info/me.json` over and over again, as fast as it can *at least* when the user does not actually have a name. Deploying this could therefor actually lead to an unwanted DOS attack:


https://github.com/elan-ev/opencast-studio/assets/1008395/36d9d3f0-49e1-407b-a963-270e687b0da0

At least the latter error should be addressed

**Comment by narickmann** [comment]
Studio should not load 'info/me.json' over and over again.

> If Opencast is unreachable, Studio shows an error. Once the connection can be established, the error disappears and the event can be uploaded. This code seems to cause a stack trace and seems to be unable to recover

This should (hopefully) also been fixed. Not sure how to test it.

File: src/steps/finish/upload.tsx [hunk 1]
```diff
@@ -184,9 +184,16 @@ const UploadForm: React.FC<UploadFormProps> = ({ handleUpload }) => {
     }
   }
 
-  // If the user has not yet changed the value of the field and the last used
-  // presenter name is used in local storage, use that.
+  // If presenter field is empty use the user.name from `/info/me.json` otherwise
+  // use value from local storage
```
Comment (LukasKalbertodt) [inline src/steps/finish/upload.tsx:188]:
```suggestion
  // If presenter field is empty, use the user.name from `/info/me.json`. Otherwise
  // use value from local storage.
```

But apart from that comma usage, that's an interesting question: is the info from `info/me.json` really more important than the one in local storage? If someone wants a different presenter name than the `info/me.json` name, then having to adjust that every time is annoying I suppose. So I would argue for prioritizing local storage.

File: src/opencast.tsx [hunk 2]
```diff
@@ -262,6 +262,24 @@ export class Opencast {
...
+      if (request) {
+        return request.user.name as string;
+      }
+    } catch (e) {
+      // If it's not our own error, rethrow it.
+      if (!(e instanceof RequestError)) {
+        throw e;
+      }
+    }
+    return null;
```
Comment (LukasKalbertodt) [inline src/opencast.tsx:280]: I would use `this.#currentUser` here. That's the result of `info/me.json` the last time the endpoint was requested. This means that we don't have a username if the endpoint was not requested yet at all, e.g. if there are no opencast connection settings set yet. But I think that only happens very rarely and shouldn't be a problem. Not having to wait for a request is worth it, I think.

Also, `this#currentUser` is `unknown`. So for typescript to accept this method, you have to have a bunch of `if (!("user" in this.#currentUser)) { return null; }` checks. If in doubt, feel free to ask me how to do this best.

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
I can confirm that both problems found by Lars seem to be fixed now!

File: src/steps/finish/upload.tsx [hunk 3]
```diff
@@ -159,18 +159,27 @@ type UploadFormProps = {
...
   const settingsManager = useSettingsManager();
   const { title, presenter, upload: uploadState, recordings } = useStudioState();
-  const presenterValue = presenter || window.localStorage.getItem(LAST_PRESENTER_KEY) || "";
+  const presenterValue = presenter
+    || window.localStorage.getItem(LAST_PRESENTER_KEY)
+    || autofillPresenter
+      .map(source => match(source, {
+        "opencast": () => opencast.getUsername(),
+      }))
+      .filter(Boolean)[0]
```
Comment (LukasKalbertodt) [inline src/steps/finish/upload.tsx:181]:
```suggestion
      .find(Boolean)
```

?

File: src/steps/finish/upload.tsx [hunk 4]
```diff
@@ -159,18 +159,27 @@ type UploadFormProps = {
...
 
 const UploadForm: React.FC<UploadFormProps> = ({ handleUpload }) => {
+  const uploadSettings = useSettings().upload ?? {};
   const {
     titleField = "required",
     presenterField = "required",
     seriesField = "optional",
-  } = useSettings().upload ?? {};
+    autofillPresenter = [],
+  } = uploadSettings;
```
Comment (LukasKalbertodt) [inline src/steps/finish/upload.tsx:168]: Why did you pull `uploadSettings` into a separate variable?

File: src/steps/finish/upload.tsx [hunk 5]
```diff
@@ -205,14 +214,6 @@ const UploadForm: React.FC<UploadFormProps> = ({ handleUpload }) => {
...
   }
 
-  // If the user has not yet changed the value of the field and the last used
-  // presenter name is used in local storage, use that.
-  useEffect(() => {
-    if (presenterValue !== presenter) {
-      dispatch({ type: "UPDATE_PRESENTER", value: presenterValue });
-    }
-  });
-
```
Comment (LukasKalbertodt) [inline src/steps/finish/upload.tsx:215]: Wait, why was this deleted?

**Review by LukasKalbertodt — COMMENTED** [review]
Thanks for trying to get this PR done! Some comments, but just minor things.

File: src/opencast.tsx [hunk 6]
```diff
@@ -288,6 +288,32 @@ export class Opencast {
...
+    }
+    if (!(typeof this.#currentUser === "object")) {
+      return null;
+    }
+    if (!("user" in this.#currentUser)) {
+      return null;
+    }
+    if (!this.#currentUser.user) {
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (JulianKniephoff) [reply src/opencast.tsx:304]: The `null` check? No, `typeof null === "object"`.

File: src/opencast.tsx [hunk 7]
```diff
@@ -288,6 +288,32 @@ export class Opencast {
     return await this.jsonRequest("info/me.json");
   }
 
+  /** Returns the value from user.name from the `/info/me.json` endpoint. */
+  getUsername(): string | null {
+    if (!(this.#currentUser)) {
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (JulianKniephoff) [reply src/opencast.tsx:295]: See above.

File: src/steps/finish/upload.tsx (hunk 4 above)
Comment (JulianKniephoff) [reply src/steps/finish/upload.tsx:168]: Remnant of an earlier version. 🙈

File: CONFIGURATION.md [hunk 8]
```diff
@@ -106,6 +106,20 @@ further below for information on that.
...
 
+# Whether and if so how to fill the presenter name in the upload step.
+# This is a list of sources for potential presenter names to suggest,
+# in order of descending preference.
+# If it is empty, nothing is suggested.
+# Note that right now there is only one such source,
+# and that manual changes to the presenter form field
+# will be persisted in the users' `localStorage`,
+# and that stored value will always be preferred
+# over the generated suggestion.
```
[the remark this answers is not in the document]
Comment (JulianKniephoff) [reply CONFIGURATION.md:117]: sure 🤷‍♀️

File: src/opencast.tsx [hunk 9]
```diff
@@ -288,6 +288,32 @@ export class Opencast {
...
+    if (!(typeof this.#currentUser.user === "object")) {
+      return null;
+    }
+    if (!("name" in this.#currentUser.user)) {
+      return null;
+    }
+    if (!(typeof this.#currentUser.user.name === "string")) {
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (JulianKniephoff) [reply src/opencast.tsx:313]: I hate that kind of indentation. But I also don't care enough to argue. 😛

File: src/steps/finish/upload.tsx (hunk 5 above)
Comment (JulianKniephoff) [reply src/steps/finish/upload.tsx]: Because I was overly eager while cleaning up. 😅 It's back now, but I took the liberty of touching it up a bit.

File: src/opencast.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply src/opencast.tsx]: Oh ye wupps

**Review by LukasKalbertodt — APPROVED** [review]
Seems to work!
